### PR TITLE
Add citation-label-format syntax

### PR DESCRIPTION
**Pull request by bwiernik**
## Description

Add syntax for specifying citation label format. Based on the biblatex label syntax: http://mirrors.ibiblio.org/CTAN/macros/latex/contrib/biblatex/doc/biblatex.pdf

It adds a `cs:citation-label-format` element.

It can take either `label-type` attribute to specify one of two prespecified formats or a combination of options attributes and rendering elements to generate a custom format.

Fixes https://github.com/citation-style-language/schema/issues/295

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

### CODE REVIEW

**Comment by denismaier** [comment]
Thanks for this. Looks like a well-crafted PR.
Can you perhaps some examples to show how the custom formats work?

Will this allow label formats like "Doe 2020"?

**Comment by bwiernik** [comment]
Here are a few examples:

Data:
>  1. (Smith, 1998)
>  1. (Doe, 2020)
>  1. (Smith, Jones, & Wilson, 2010)
>  1. (Smith, Johnson, & Wendel, 2008)
>  1. (Smith & Le, 2007)
>  1. (Smith & Lenore, 2008)
>  1. (Smith, Larson, Nkomo, & Haverford, 2015)

### Style 1


Code: 
```
<citation-label-format width="1" width-single="3" disambiguate-add-characters="true" characters-disambiguation-rule="by-position">
  <names variable="author" form="short"/>
  <date variable="issued" form="short" date-parts="year"/>
</citation-label-format>
```

Result:
>  1. Smi98
>  1. Doe20
>  1. SJonWi10
>  1. SJohWe08
>  1. SLe07
>  1. SLen08
>  1. SLNH15


### Style 2

Code: 
```
<citation-label-format width="1" width-single="3" disambiguate-add-characters="true" characters-disambiguation-rule="by-position" fixed-width="true" pad-character="x">
  <names variable="author" form="short"/>
  <date variable="issued" form="short" date-parts="year"/>
</citation-label-format>
```

Result:
>  1. Smi98
>  1. Doe20
>  1. SJonWi10
>  1. SJohWe08
>  1. SLex07
>  1. SLen08
>  1. SLarNH15


### Style 3

Code: 
```
<citation-label-format width="1" width-single="3" disambiguate-add-characters="true" characters-disambiguation-rule="by-cite">
  <names variable="author" form="short"/>
  <date variable="issued" form="short" date-parts="year"/>
</citation-label-format>
```

Result:
>  1. Smi98
>  1. Doe20
>  1. SJW10
>  1. SJW08
>  1. SL07
>  1. SL08
>  1. SLNH15


### Style 4

Code: 
```
<citation-label-format width="1" width-single="3" fixed-width="3">
  <names variable="author" form="short"/>
  <date variable="issued" form="short" date-parts="year"/>
</citation-label-format>
```

Result:
>  1. Smi98
>  1. Doe20
>  1. SJW10
>  1. SJW08
>  1. SmL07
>  1. SmL08
>  1. SLN15


### Style 5

Code: 
```
<citation-label-format width="1" width-single="3" disambiguate-add-characters="true" characters-disambiguation-rule="by-position" et-al-min="4" et-al-use-first="3" et-al-form="symbol">
  <names variable="author" form="short"/>
  <date variable="issued" form="short" date-parts="year"/>
</citation-label-format>
```

Result:
>  1. Smi98
>  1. Doe20
>  1. SJonWi10
>  1. SJohWe08
>  1. SLe07
>  1. SLen08
>  1. SLN+15


### Style 6

Code: 
```
<citation-label-format delimiter=" " et-al-min="3" et-al-use-first="1" et-al-form="long">
  <names variable="author" form="short" delimiter=", "/>
  <date variable="issued" form="short" date-parts="year"/>
</citation-label-format>
```

Result:
>  1. Smith 1998
>  1. Doe 2020
>  1. Smith et al. 2010
>  1. Smith et al. 2008
>  1. Smith, Le 2007
>  1. Smith, Lenore 2008
>  1. Smith et al. 2015

File: schemas/styles/csl-citation-label-format.rnc [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  style.citation-label-format =
+    
+    ## Specify the format for automatically-generated citation-labels.
+    ## If unspecified, citation-labels are not automatically generated.
+    element cs:citation-label-format {
+      (
+       ## Specify a predefined label style
+       attribute label-style {
+         
+         ## Style used in citeproc-js for CLS 1.0.* styles
```
Comment (bdarcus) [inline schemas/styles/csl-citation-label-format.rnc:16]: It's more important on the docstrings to include an example so someone can easily understand what the format is.

Comment (bdarcus) [inline schemas/styles/csl-citation-label-format.rnc:16]: It's more important on the docstrings to include an example so someone can easily understand what the format is.

File: schemas/styles/csl-citation-label-format.rnc [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+         "citeproc"
+         | 
+           ## Default style used in biblatex
```
Comment (bdarcus) [inline schemas/styles/csl-citation-label-format.rnc:19]: Same here.

File: schemas/styles/csl-citation-label-format.rnc [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    ## Default values for inheritable attributes in citation-label-format
+    ## [ a:defaultValue = "" ] names-delimiter
+    ## [ a:defaultValue = "" ] name-delimiter
+    ## [ a:defaultValue = "short" ] name-form
+    names-inheritable-options,
+    name-inheritable-options,
+    
+    ## Abbreviate non-date variables by extracting the first N characters.
+    ## For list variables (e.g. names), extract chacters from each element.
+    attribute width { xsd:integer }?,
```
Comment (bdarcus) [inline schemas/styles/csl-citation-label-format.rnc:42]: If this is optional, and there's no default, what happens if it's missing? The complete name is rendered?

File: schemas/styles/csl-data.rnc [hunk 4]
```diff
@@ -1,14 +1,20 @@
+namespace a = "http://relaxng.org/ns/compatibility/annotations/1.0"
 namespace dc = "http://purl.org/dc/elements/1.1/"
 
 dc:title [ "Citation Style Language Data" ]
 dc:creator [ "Bruce D'Arcus" ]
-dc:rights [ "Copyright 2009-2018 Citation Style Language and contributors" ]
```
Comment (bdarcus) [inline schemas/styles/csl-data.rnc:5]: Why are there all these formatting/whitespace changes?

Comment (bwiernik) [reply schemas/styles/csl-data.rnc:5]: It's what trang is changing when I run it.

File: schemas/styles/csl-citation-label-format.rnc (hunk 3 above)
Comment (bwiernik) [reply schemas/styles/csl-citation-label-format.rnc:42]: Yes. If width is empty, no abbreviation is applied.

File: schemas/styles/csl-data.rnc (hunk 4 above)
Comment (bdarcus) [reply schemas/styles/csl-data.rnc:5]: Interesting. I guess the question is how the repo ended up with the funky formatting.

Oh well, not germane to the PR.

**Pull request merged**